src/PowerShellEditorServices.Host/CodeLens/PesterCodeLensProvider.cs

```diff
@@ -33,6 +40,38 @@ public PesterCodeLensProvider(EditorSession editorSession)
         {
             _editorSession = editorSession;
             _symbolProvider = new PesterDocumentSymbolProvider();
+
+            DeterminePesterVersion();
+        }
+
+        /// <summary>
+        /// Used to determine the value of <see cref="_pesterV4_6_0_OrHigherAvailable"/> as a background task.
+        /// </summary>
+        private void DeterminePesterVersion()
```

Comment:
I'm worried that this is a task in disguise. As a `void` method around a `Task.Run` invocation, it either blocks the caller or it's a race condition around the `_pesterV4_6_0_OrHigherAvailable` variable.

This strikes me as much better implemented as a `Task<bool>`:

```csharp
private async Task<bool> HasPester460Async()
{
    var pesterModSpec = new Hashtable()
    {
        { "ModuleName", "Pester" },
        { "ModuleVersion", new Version(4, 6) }
    };

    var gmoListPesterCommand = new PSCommand()
        .AddCommand("Microsoft.Powershell.Core\\Get-Module")
            .AddParameter("ListAvailable")
            .AddParameter("FullyQualifiedName", pesterModSpec);

    IEnumerable<PSModuleInfo> modules = await _powerShellContext.ExecuteCommandAsync<PSModuleInfo>(
        gmoListPesterCommand,
        sendOutputToHost: false,
        sendErrorToHost: false);

    return modules != null && modules.Any();
}
```

Reply:
👍 to @rjmholt's comment. At the very least it would be nice to store the result as a task to avoid race conditions.

Reply:
The other thing is that making this an async method makes it much harder to call from the constructor. I think that's a good thing, because I'm not a big fan of complex constructor logic, and tasks in constructors make me uneasy about PSES' already bad thread situation.

Instead, the best solution to me feels like a [lazy async](https://blog.stephencleary.com/2012/08/asynchronous-lazy-initialization.html) field created in the constructor around the method above

Reply:
The idea was that the default value of a boolean is false, which means the behavior will be the old one, when the async task finishes, then the value can be true and only then the system would switch over to the new improved behaviour, therefore I saw this race condition as by design. I am not a fan of constructor work either because it can make unit testing work hard, I could put the work in an `Initialize()` method instead if you prefer that.
I suggest to rather put a blocking wait when `GetPesterLens()` gets called to wait until the async task has reported the result if you want to completely avoid existing behaviour if someone clicks the Run Tests button too quickly immediately after opening the editor on a slow machine. I would then try to write the code that there is no block if the describe block string is not interpolated.
Would you be happy with that?